### ISSUE TITLE
multi pumping

### ISSUE THREAD

**Issue by benjaminkreis** [fastmachinelearning/hls4ml#99]
Starting a new issue so we can discuss multi pumping 

Following up on @zhenbinwu's presentation today, I tried running our 1 hidden layer example with the multipliers using LUT-based cores instead of DSP-based cores (ie `#pragma HLS RESOURCE variable=my_var core=Mul_LUT`)

Using Vivado 2018.2 and default hls4ml options from the head, this is what I see:

|          | Latency          | DSP |  FF  | LUT |
| ------------- |:-------------:| -----:| -----:| -----:|
| DSP multipliers |12 |  305 | 9798 | 11819 |
| LUT multipliers | 11  |  0 | 9127 | 137371 |

I think this is more like what we expected to see -- that the DSPs go to zero, and we use way more LUTs.

The number of LUTs per multiplication roughly makes sense I think.  If you take the excess LUTs and divide by the number of DSPs that were used, you get ~411 LUTs per multiplication.  If you create a 18bit x 18bit LUT-based multiplier IP core in regular Vivado, you get 365 or 401 LUTs per multiplication depending on the optimization option (non constant coefficient).

It will be interesting to find out how this multi pumping code uses so few!

**Comment by zhenbinwu** [comment]
yeah, I am curious about this pumping builtin function too. I am worrying they might be too good to be true. It also use much less FF as well, only 1405! 

What is the precision you used here?

**Comment by benjaminkreis** [comment]
Right!

I used <18,8>

**Comment by zhenbinwu** [comment]
Can you test one with <16, 4>? That is the one I am using for the slides. I am also curious how will the FF scale with number of bits.
